[jit] Fix getConstant

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29220 [quant][graphmode] Enable foldbn tests
* #29219 [jit] Expose class constant through `attr` and `setattr` in object
* #30787 [jit] Add unsafeRemoveConstant for ClassType
* **#31012 [jit] Fix getConstant**
* #29218 [jit] Preserve constant from ConcreteModuleType to ClassType

Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

jerryzh168 committed:
[jit] Fix getConstant

Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

suo reviewed (changes requested):
Same note as the other PR: we need `findConstant`.

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 7d42881:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7d42881a71542a6c4e53c0cd3305af3e76eb888a).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3919484) binary_linux_manywheel_3_7mu_cpu_devtoolset7_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72954613))
```
Dec 12 05:57:03 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o
Dec 12 05:57:04 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o
Dec 12 05:57:04 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o
Dec 12 05:57:04 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o
Dec 12 05:57:05 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/fuse_linear.cpp.o
Dec 12 05:57:05 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o
Dec 12 05:57:05 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/fuser/interface.cpp.o
Dec 12 05:57:06 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_prim_ops.cpp.o
Dec 12 05:57:06 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_string_ops.cpp.o
Dec 12 05:57:06 /pytorch/torch/csrc/jit/passes/python_print.cpp: In member function ‘void torch::jit::PythonPrintImpl::printModuleMetadata(const ClassTypePtr&)’:
Dec 12 05:57:06 /pytorch/torch/csrc/jit/passes/python_print.cpp:1263:53: error: ‘struct c10::IValue’ has no member named ‘value’; did you mean ‘IValue’?
Dec 12 05:57:06        const auto& v = moduleType->getConstant(name).value();
Dec 12 05:57:06                                                      ^~~~~
Dec 12 05:57:06                                                      IValue
Dec 12 05:57:07 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/register_special_ops.cpp.o
Dec 12 05:57:08 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/scope.cpp.o
Dec 12 05:57:10 gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o] Error 1
Dec 12 05:57:10 gmake[2]: *** Waiting for unfinished jobs....
Dec 12 05:57:31 gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2
Dec 12 05:57:31 gmake: *** [all] Error 2
Dec 12 05:57:31 Traceback (most recent call last):
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -112,8 +112,35 @@ size_t ClassType::addConstant(
-c10::optional<IValue> ClassType::getConstant(const std::string& name) const {
-  TORCH_CHECK(constantNames_.size() == constantValues_.size());
+IValue ClassType::getConstant(const std::string& name) const {
+  TORCH_INTERNAL_ASSERT(constantNames_.size() == constantValues_.size());
```
Same comment as the other one; let's re-use `findConstant` instead of duplicating the code.

suo commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -112,8 +112,35 @@ size_t ClassType::addConstant(
+
+IValue ClassType::getConstant(size_t slot) const {
+  TORCH_INTERNAL_ASSERT(constantNames_.size() == constantValues_.size());
+  TORCH_CHECK(slot < constantValues_.size());
```
Let's add an error message like the one in the other `getConstant`

suo reviewed (approved):
Accepting but please address inline comments!

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getConstant"


Summary:
- getConstant should throw when the item is not found
- add another getConstant which takes slot index as argument

Test Plan:
test_class_type.cpp

Reviewers:
suo
Subscribers:

Tasks:

Tags:

Differential Revision: [D18898418](https://our.internmc.facebook.com/intern/diff/D18898418)

[ghstack-poisoned]
